### ISSUE TITLE
Texture memory leak in some games

### ISSUE THREAD

**Issue by loganmc10** [gonetz/GLideN64#1015]
I believe I am experiencing a memory leak in some games, the 2 games I consistently have this problem with are GoldenEye and Worms Armageddon, other games work fine, even if I play them for a long time.

I have tried to adjust the texture cache size, it seems to make no difference. The Raspberry Pi GPU has ~400MB of memory.

When I play Worms Armageddon, the GPU memory on my Raspberry Pi slowly fills up (about 1MB/sec or two). It only happens once I'm "in-game", not in the menu. Once the problem starts, the textures start to swap with each other and everything looks all garbled, here are some pictures from GoldenEye showing what happens (sorry for the cell phone pics, once this happens the program I use to take screenshots no longer works).

![vrhkr3e](https://cloud.githubusercontent.com/assets/848146/15829537/6d4163fe-2bd1-11e6-833c-555b0b6ecea8.jpg)
![46cbfle](https://cloud.githubusercontent.com/assets/848146/15829541/72926b78-2bd1-11e6-938c-eb1b6fec3a23.jpg)

**Comment by loganmc10** [comment]
Perhaps this is somewhat related to this comment? https://github.com/gonetz/GLideN64/pull/745#issuecomment-146402996

I'm not really sure

**Comment by loganmc10** [comment]
By the way, this happens before my device runs out of memory. This is what my GPU details show when the problem happens:

(In this case I gave the GPU 680M of memory, the problem occurred when the free memory reached 342M). When everything is good, the "alloc_fails" number is 0, when the problem begins, alloc_fails starts to climb, and the textures start to get messed up. My texture cache is at the default of 500MB

```
Relocatable heap version 4 found at 0x13400000
total space allocated is 680M, with 678M relocatable, 2.3M legacy and 0 offline
1 legacy blocks of size 2359296

small_allocs                  : 239
allocs                        : 32813
alloc_fails                   : 94860
legacy_block_fails            : 0
compactions                   : 0
discard_compactions           : 0
aggressive_compactions        : 0
aggressive_compaction_waits   : 0
aggressive_compaction_timeouts: 0
locks                         : 0
small_locks                   : 0

free list at 0x3d97d6a0
342M free memory in 17 free block(s)
largest free block is 342M bytes
```

**Comment by loganmc10** [comment]
Ok, so this happens 100% of the time when "allocs" reaches ~32800 (I assume the actual "problem number" is 32767).

When I stop mupen64plus and restart it, everything works fine, and the "allocs" number keeps climbing properly.

**Comment by loganmc10** [comment]
So according to this: https://github.com/raspberrypi/firmware/issues/611 it does sound like the Pi supports a maximum of 32768 allocations. Should GlideN64 be doing this? Would it ever end? It seems like Worms Armageddon is just endlessly allocating new textures.

Also they mentioned "All gl calls must come from the same thread. I have seen a bug where texture deletes were being done from a different thread. That violates the GL API, and on the Pi the texture deletes will fail."

Is that the case here?

**Comment by loganmc10** [comment]
Could this be solved by telling TextureCache::_checkCacheSize() to delete from the cache if the number of items in the cache is over 32000 (just on the Raspberry Pi)?

**Comment by gonetz** [comment]
> Also they mentioned "All gl calls must come from the same thread. I have seen a bug where texture deletes were being done from a different thread. That violates the GL API, and on the Pi the texture deletes will fail."
> 
> Is that the case here?
> 

Hardly. Project64 can call plugin's API functions from different threads, thus Zilmar specs build uses separate thread for all GL calls. mupen64plus calls all API functions from the same thread. There is a chance that it is broken on Pi, but I don't believe it.

> Could this be solved by telling TextureCache::_checkCacheSize() to delete from the cache if the number of items in the cache is over 32000 (just on the Raspberry Pi)?

Well, yes it can be solved that way.

**Comment by fzurita** [comment]
A bigger question is why are there so many textures? It almost seems like the same textures are getting cached over and over.... in that case the cache seems useless seems nothing is being found in the cache.

**Comment by gonetz** [comment]
It needs investigation. You may try to prove that "same textures are getting cached over and over":
for each new texture calculate checksum of GL texture data and put it in a list. Try to find crc of new GL textures in that list. If you will really find duplicates it will prove your hypothesis.

**Comment by loganmc10** [comment]
This has been fixed with this commit:
https://github.com/gonetz/GLideN64/commit/8c632ca0adcd085c2888bc77479ae0ece7316441

Do you want to leave this open until someone looks into the duplicate thing? I'm happy with how it is now.

**Comment by AmbientMalice** [comment]
It's likely related to https://github.com/gonetz/GLideN64/issues/997 which is kind of a serious problem.

**Comment by matthewharvey** [comment]
When I ran into this before, I figured that the game was creating the textures on the fly for single use. It certainly seemed that way in Beetle Adventure Racing, and after looking at the dumped textures they seemed to be creating the in-game map on the fly.
That would fill up the texture cache, so the only thing necessary for the RPi would be to make the texture cache size (one of the configuration options)  is really small so that there were never more than 15000 textures in the cache. I'm not opposed to the solution that has been put in place, but a nice comment block explaining why it is there would be helpful for the future.
I'm not sure about any relation to 997, but it might be worth a check to dump out the texture cache and ensure there aren't any duplicate crcs in there.

**Comment by loganmc10** [comment]
Lowering the cache size (in MB) isn't a guaranteed solution since you don't know how big each texture is, the only real way to guarantee you won't have an issue is to limit the number of items in the cache

**Comment by loganmc10** [comment]
So I confirmed that each CRC is unique inside the texture cache (using Worms Armageddon, since it is one of these games that seems to endlessly generate textures) by adding ```LOG(LOG_WARNING, "%i\n",_crc32);``` to TextureCache::_addTexture(u32 _crc32) and using a script to make sure there were no duplicates.

I assumed there wouldn't be, since the logic is there already to not add a texture with a matching CRC, however I would be interested to know if the dumped textures themselves have any duplicates (checking the CRC/md5sum of each dumped texture itself).

**Comment by loganmc10** [comment]
So I did a texture dump on Worms Armageddon and checked the md5sum of the textures that were dumped, there were lots of duplicates:

https://github.com/gonetz/GLideN64/issues/997#issuecomment-224926372

**Comment by gonetz** [comment]
>  by adding LOG(LOG_WARNING, "%i\n",_crc32); to TextureCache::_addTexture(u32 _crc32)

You dig in a wrong place. Of course, all items in texture cache have unique CRC - texture cache guarantees that. However, unique CRC of texture cache item does not guarantee unique content of its texture. To check that you need to calculate checksum of texture data prepared in TextureCache::_load, namely pDest provided to glTexImage2D.

**Comment by loganmc10** [comment]
Ya I figured the CRC's would all be unique. Just to make sure I'm doing this right:

If I check the value of pDest right before this line:
https://github.com/gonetz/GLideN64/blob/master/src/Textures.cpp#L1156

Each value should be unique?

**Comment by gonetz** [comment]
Value of pDest will give you nothing. You need to compare its content. That is calculate checksum of the data pDest points to: crc = CRC_Calculate(0xFFFFFFFF, pDest, _pTexture->textureBytes)

**Comment by loganmc10** [comment]
Maybe this is beyond my understanding, doesn't ```*pDest``` give the value/contents of pDest? It is a pointer to a 4-byte unsigned integer, what is the purpose of calculating a checksum on that?

**Comment by fzurita** [comment]
Yeah, pointers are tricky in C++. pDest is a pointer to an array of data. *pDest gives you only the first item of the array. To get the value of more than one item, you would have to do something like this:

```
int firstItem = *pDest;
pDest++;
int secondItem = *pDest
```

and so on.

**Comment by loganmc10** [comment]
oh ok that makes sense, and I assume that ```_pTexture->textureBytes``` gives me the length of the array that starts at that pointer? I can work with that.

**Comment by fzurita** [comment]
Yes, except in bytes. So the length of the pDest array is _pTexture->textureBytes/4

**Comment by loganmc10** [comment]
Ok so I added this to the TextureCache::_load function:

```
for( int my = 0; my < _pTexture->textureBytes/4 ; my = my + 1 )
       {
               LOG(LOG_WARNING, "%u ", pDest[my]);
       }
LOG(LOG_WARNING,"\n");
```
That prints out lines that look something like this:
```0 0 1934295040 2206434253 1934328707 0 0 0 0 1640300544 2336981891 2198047563 2336981891 33667 0 0 0 2211250176 2198045571 2206434179 2198045571 1783204683 0 0 0 2206429955 2206434051 2211283917 2198045571 1934330699 0 0 0 2206434179 2206434179 2211283917 2198045571 1929610115 0 0 0 2211279691 2206433923 2206434179 2198045571 35799 0 0 0 2206400512 2198045571 2198045443 2206434051 21001 0 0 0 0 2211279691 2211283843 29515 0 0 0```

Then I use python to look at each line in gliden64.log and file duplicates, does this all look right? (It finds lots of duplicates by the way)

**Comment by fzurita** [comment]
That looks like the correct way to check to me.

**Comment by loganmc10** [comment]
Another way I tried checking is using the ```boost::hash_range()``` function:

```
size_t output = boost::hash_range(pDest, pDest + (_pTexture->textureBytes/4)-1); LOG(LOG_WARNING,"%zu\n",output);
```

I believe I'm doing that right? That prints out a hash of pDest if I understand it correctly.

After letting Worms Armageddon run for a minute or two, it recorded 2100 hashes (from the _load() function), and 141 of those had at least 1 duplicate.

**Comment by gonetz** [comment]
Correct way is to calc and print data checksum: crc = CRC_Calculate(0xFFFFFFFF, pDest, _pTexture->textureBytes)

**Comment by loganmc10** [comment]
What would be the purpose of testing that? That is what it's already doing isn't it? Could these duplicates be happening because the current CRC is insufficient?

**Comment by gonetz** [comment]
Cr in texture cache is checksum of data in rdram. If you want to find duplicates in video memory, you need to compare texture data. Texture data obviously differs from rdram data.

**Comment by loganmc10** [comment]
Ok so I have tested this 3 different ways: dumping the value of pDest in a for loop and comparing the output, hashing the value with boost::hash_range and comparing the hashes, and using ```CRC_Calculate(0xFFFFFFFF, pDest, _pTexture->textureBytes)``` and comparing that value, here are my results:

I used a save state right at the beginning of a Worms Armageddon match and let the game play out (I didn't push any buttons) until _load() had processed around 5500 textures.

Dumping the value of pDest:
```
Duplicates:
2024 / 5923
34.17%

Unique:
3899
```

boost::hash_range:
```
Duplicates:
1908 / 5204
36.66%

Unique:
3296
```

CRC_Calculate:
```
Duplicates:
1914 / 5271
36.31%

Unique:
3357
```

The "duplicates" field counts the "original" texture, so there is marginally fewer duplicates than it says there, but it seems to me that no matter how I calculate it, using pDest, there are duplicate textures being loaded into the texture cache, maybe it would be helpful if someone else could confirm these findings?

**Comment by gonetz** [comment]
Thanks! This means that checksum of rdram data is too strict: it considers that data is different, while actually it can be the same. I'll check it, someday. Or you may try to detect, when duplicated t

**Comment by loganmc10** [comment]
Ok so I looked into this and compared textures that had the same output from ```CRC_Calculate(0xFFFFFFFF, pDest, _pTexture->textureBytes)```.

These are the ```TextureParams``` fields that can have different values even though the pDest CRC value is identical:

**clampWidth**:
This was the biggest one, this value could vary widely even though the pDest CRC was the same, for example, a CRC of 3530616496 had these unique values for clampWidth :
3530616496 ['8', '1016', '1017', '1022', '1021', '1018']
There were lots of these.

**clampHeight**:
After running for a while, there were only 3 duplicate textures that had multiple values for this, and it was only off by one (for instance, the same CRC had 1022 and 1023 for clampHeight)

**clampS**:
This happened lots of the time as well, clampS could be 0 or 1 for the same CRC value.

Every other ```TextureParams``` value was good (there was only 1 value per pDest CRC)

**Comment by loganmc10** [comment]
hmm but now that I look at it, I don't think the clamp* values are used in the CRC calculation, I'll have to keep looking...

**Comment by loganmc10** [comment]
I confirmed this by removing ```clampWidth``` and ```clampHeight``` from ```TextureParams``` (I left clampS in there).

Here are my results:
```
Duplicates:
190/5198
3.66%
```

Honestly the image looks the same (Worms Armageddon), although I will need to spend some more time with it to be sure.

Removing clampS and clampT drops the duplicate numbet down to 0, although if I play Super Smash Bros, it generates about 20% duplicates with those settings, so I'm going to look into that now...

Even with clampS and clampT taken out of ``TextureParams`` the picture looks normal

**Comment by loganmc10** [comment]
Super Smash Bros is all over the place, there are a handful of instances where almost every item in ```TextureParams``` is different between 2 matching pDest's.

But actually what is most interesting is that in most cases, the pDest CRC and every item in ```TextureParams``` is identical between duplicates. I'm not really sure what to make of that.

**Comment by loganmc10** [comment]
So after looking into ```_calculateCRC()``` it is sometimes this line:
https://github.com/gonetz/GLideN64/blob/master/src/Textures.cpp#L1184

and sometimes this line:
https://github.com/gonetz/GLideN64/blob/master/src/Textures.cpp#L1193

That "throws off" the CRC calculation, depending on the texture.

Or it is sometimes this line:
https://github.com/gonetz/GLideN64/blob/master/src/Textures.cpp#L1200

Because of the issues with some of the ```TextureParams``` like I have been mentioning.

**Comment by gonetz** [comment]
Great findings, thanks!

> 3530616496 ['8', '1016', '1017', '1022', '1021', '1018']

Is it from Worms Armageddon? Such values for clampWidth are meaningless for normal textures. I guess, that is parameters of background texture.

> So after looking into _calculateCRC() it is sometimes this line:

The first CRC_Calculate calculates checksum of texture data in N64 texture memory. The second  CRC_Calculate is used for paletted texture formats: if palette is different, output texture also will be different. The third CRC_Calculate calculates checksum of texture parameters.  This can be redundant. 

> But actually what is most interesting is that in most cases, the pDest CRC and every item in TextureParams is identical between duplicates.

Yes, this is interesting. This must not be. Super Smash Bros you say?

**Comment by loganmc10** [comment]
Yes those clampWidth numbers are from Worms Armageddon, I don't really know what texture produced them, but I got them from the _load() function (not _loadBackground()). 

And yes, I saw many textures that had identical pDest CRCs and identical TextureParams, the thing that crested a different CTC was either that first CRC_Calculate that is the checksum of the N64 texture memory or the CRC_Calculate for the paletted texture formats.

That happened in Super Smash and GoldenEye

**Comment by loganmc10** [comment]
I've been playing around with different ways to calculate the CRC and I can't seem to get rid of this problem. I think I've come up with a possible solution:

When a texture is loaded (using ```_load()``` or ```_loadBackground()```) record the pDest CRC value along with the name of the texture.

When a future texture is loaded, check to see if there is a matching pDest CRC out there, if there is, use glTexSubImage2D to update that texture, instead of creating a new one. Based on my reading that should be a much faster operation.

Does that sound reasonable? (I just want to check before I spend the time coding it)

**Comment by loganmc10** [comment]
I might even try not doing glTexSubImage2D, just bind to the "original" texture and use that.

**Comment by LegendOfDragoon** [comment]
You mind testing SD Hiryuu no Ken Densetsu? I think that game also has some texture problems.

**Comment by loganmc10** [comment]
@LegendOfDragoon what do you mean by texture problems?

**Comment by LegendOfDragoon** [comment]
![image](https://cloud.githubusercontent.com/assets/10492645/16465200/5762ed22-3df2-11e6-8b32-1e1959099625.png)

Here's what the menu is like, in a build that's from a few weeks ago. It's very slow. Last time I checked, it was some problem with texture loading / crc issue.

**Comment by gonetz** [comment]
I've made some fixes for texture cache. Could you update current status of the problem: which games still have texture cache overflow problem? I know only #795, but it is very special case.

**Comment by LegendOfDragoon** [comment]
Where can I download the latest build?

**Comment by loganmc10** [comment]
This seems like it is fixed, Worms Armageddon now reports no duplicate textures (checking pDest CRC's). Smash Bros now reports about 7% duplicates, it was about 20% before. Goldeneye used to report up to 80% duplicates while playing the opening title sequence, now it's down to about 20%, all big improvements. I'm happy with how it is now. It's probably not ever going to be perfect, but this is a huge improvement. I'm happy to close this issue if you are @gonetz

**Comment by gonetz** [comment]
Thanks for the statistics. I close this ticket.

I investigated some cases with duplicates, which appears after my fixes. I came to conclusion that checksum algorithm worked correct in these cases. Most duplicates are among color indexed textures. Input tiles may have different patterns and thus different crc, but if they use the same palette with one-two colors in it, resulted textures can be the same. It also possible that the same texture image is encoded differently, but when it is decoded to RGBA format from different encodings, result is the same.
